### PR TITLE
TotW: Branch, Review, and Learn

### PR DESCRIPTION
**Pull request by d33bs**
This PR adds a new tip of the week post titled "Branch, Review, and Learn" covering git best practices with a focus on branching, pull/merge request reviews and merging strategies (especially in the context of Github). Given this is a space with many ideas and tools, and there are many existing guides on performing the technical aspects, I tried to keep the focus on conceptual elements that make sense based on my own experiences (also providing a caveat near the beginning about deferring to existing developer culture, wherever the reader may be). This candidly is a very meta PR! 🙂 

Thank you in advance for any thoughts, comments, or feedback you may have!

### CODE REVIEW

File: _posts/2023-02-13-Branch-Review-and-Learn.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+       branch learner
+       checkout learner
+       commit id: "completed"
+       checkout main
+</pre>
+
+_An example git diagram showing learner branch based off main._
+{:.center}
+
+Following the course assignment workflow, the diagram above shows an in-progress `learner` branch based off of the `main` branch. When the `learner` branch is created, we bring into it everything we know from `main` (the course) so far in the form of [commits](https://github.com/git-guides/git-commit), or groups of changes to various files. [Branching](https://github.com/git-guides#create-a-branch) allows us to make consistent and well described changes based on what's already happened without impacting others work in the meantime.
```
Comment (falquaddoomi) [inline _posts/2023-02-13-Branch-Review-and-Learn.md:76]: First, I love these Git diagrams! I like the example as well, but I think naming the branch `learner` gets away from the idea you bring up elsewhere of using branches for small tasks. Perhaps naming it `assignment` or some other indication that it's a task, not the student's personal branch, would make things clearer.

File: _posts/2023-02-13-Branch-Review-and-Learn.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+</pre>
+<script type="module">
+  import mermaid from 'https://unpkg.com/mermaid@9/dist/mermaid.esm.min.mjs';
+  mermaid.initialize({ startOnLoad: true });
+</script>
+
+_An example course and learner assignment workflow._
+{:.center}
+
+[Git branching](https://www.atlassian.com/git/tutorials/using-branches) practices may be understood in context with similar workflows from real life. Consider a scenario within an educational course where an assignment is made available to a learner. In addition to the steps shown in the diagram above, it's important to think about _why_ this pattern is beneficial:
```
Comment (falquaddoomi) [inline _posts/2023-02-13-Branch-Review-and-Learn.md:50]: Is there a reason you're calling the person taking a course a "learner" rather than a "student"? I just feel like this text is unnecessarily formal for the intended audience, who I assume is lay people interested in using Git. I'd prefer the scenario to be explained in simpler terms, maybe something like "Consider a student taking a course, where an assignment is given to them to complete."

File: _posts/2023-02-13-Branch-Review-and-Learn.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+       checkout learner
+       commit id: "completed"
+       checkout main
+       merge learner id: "reviewed"
+</pre>
+
+_An example git diagram showing learner branch being merged with main after a review._
+{:.center}
+
+The diagram above depicts a merge from the `learner` branch to pull the changes into the `main` branch, simulating an assignment being returned for review within a course. While merges may be forced without review, it's a best practice to ask others for a [Pull Request (PR) Review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) (also known as a [Merge Request (MR)](https://docs.gitlab.com/ee/user/project/merge_requests/) on some systems). Doing this provides a chance to make revisions before code changes are "finalized" within the `main` branch.
```
Comment (falquaddoomi) [inline _posts/2023-02-13-Branch-Review-and-Learn.md:105]: I'm being more than a little pedantic here, but technically a developer issues a pull request, then the PR is reviewed (or not), and then potentially merged. IMHO it's important to identify that the pull request is its own entity, separate from the review. Maybe you could reword this as "While merges may be forced without review, it's a best practice create a Pull Request (PR) and then ask other members of your team to review it."

**Review by falquaddoomi — APPROVED** [review]
This looks good overall, and I especially like the styling of the Git diagrams. I made a few comments, but I don't consider them essential so I'm approving this for now. I could see this being an introduction to the concept of branching and pull requests, but (depending on your vision of your audience) it would be nice to see a hands-on post on specific Git commands and GitHub features you'd use to accomplish this.

File: _posts/2023-02-13-Branch-Review-and-Learn.md (hunk 2 above)
Comment (d33bs) [reply _posts/2023-02-13-Branch-Review-and-Learn.md:50]: Thanks, I see your point, and can move into student instead to simplify. I felt that the audience here might not be those interested in using git but perhaps those who are building software. Some of the content is built with the idea that they may need convincing to use additional features which may appear arbitrary or overly complex without more explanation. The "learner" use over "student" was to help avoid the connotation of hierarchy in the midst of parallels to a developer team (where there may not be a formal "teacher" so much as all being somewhat equally instructive).

File: _posts/2023-02-13-Branch-Review-and-Learn.md (hunk 3 above)
Comment (d33bs) [reply _posts/2023-02-13-Branch-Review-and-Learn.md:105]: Thanks, I like that change, I'll add that clarity into the wording here.

File: _posts/2023-02-13-Branch-Review-and-Learn.md (hunk 1 above)
Comment (d33bs) [reply _posts/2023-02-13-Branch-Review-and-Learn.md:76]: Love this, thanks! It definitely aligns well to the purpose rather than role for that lane/branch to demonstrate the significance here. I'll make changes accordingly.

**Review by vincerubinetti — APPROVED** [review]

**Comment by vincerubinetti** [comment]
Note that the `<!-- excerpt start/end -->` tags haven't been implemented yet, but they'll be in the new version of the template soon.

**Comment by vincerubinetti** [comment]
Note that setting `excerpt` in front matter also won't work yet. By default, Jekyll sets `excerpt` to be the first paragraph of the post. Since the first paragraph can be either very long or very short, I elected to use (in the post-excerpt component) the whole content of the post instead and hard-limit it to a certain amount of characters, so the displayed excerpts were more consistent sizes.

In the new version of the template I will revert to the default `excerpt` behavior and add the `<!-- -->` markers. I would just leave your posts as they are for now, or add the markers back in.

**Comment by d33bs** [comment]
Understood, thank you @vincerubinetti !
